### ISSUE TITLE
Don't use impl Connect in return types

### ISSUE THREAD

**Issue by golddranks** [fussybeaver/bollard#17]
Currently functions like `BollardDockerApi::connect_with_unix_defaults` return `Result<Docker<impl Connect>, Error>`. This has some downsides from the API viewpoint: the users of the API are unable to store the returned value in a struct, because Rust doesn't allow using `impl Connect` in structs yet. It would be better to just return a concrete type for the time being.

**Comment by fussybeaver** [comment]
Thanks for pointing this out, it's definitely a weak point in the API, which I intend to address.

I guess right now you would do something like:

```rust
struct Meta<C>
where
    C: Connect,
{
    client: Docker<C>,
}

fn main() {
    #[cfg(unix)]
    let docker = Docker::connect_with_unix_defaults().unwrap();
    #[cfg(windows)]
    let docker = Docker::connect_with_named_pipe_defaults().unwrap();

    let meta = Meta {
        client: docker,
    };

    // ...
}
```

The API returns an `impl Trait` so that you can abstract between different platforms. 

In future, I would prefer to use an "enum" instead of `Connect` as a concrete type, which may be easier to work with and looks a bit cleaner.

**Comment by golddranks** [comment]
Thanks for the reply. I manage to get by with that pattern, although it's undesirable that the "parametricity" bleeds out of the containing type.

I think that https://github.com/rust-lang/rfcs/pull/2515 is going to eventually solve this problem, making the type expressable. However, if runtime dispatch over the selected type is desirable, enums are the correct solutions and they are available right away :)

**Comment by veeg** [comment]
As a side note, I would like to expand on this problem with a concrete use-case that I am unable to fulfill right now, and discuss the alternatives.

## Use-case:
Abstract away from the user of _my_ library that a docker connection is even present.

Attempting to hide away the instantiation, for instance like below, will result in mismatch types:
```rust
use bollard::Docker;
#[cfg(unix)]
use hyperlocal::UnixConnector;
#[cfg(windows)]
use named_pipe::NamedPipeConnector;

pub struct DockerTest {
    #[cfg(unix)]
    client: Docker<UnixConnector>,
    #[cfg(windows)]
    client: Docker<NamedPipeConnector>,
}

impl DockerTest {
    pub fn local() -> DockerTest {
        #[cfg(unix)]
        let docker = Docker::connect_with_unix_defaults().unwrap();
        #[cfg(windows)]
        let docker = Docker::connect_with_named_pipe_defaults().unwrap();

        DockerTest { client: docker }
    }
}

fn main() {
    let _test = DockerTest::local();

    // test.run(|| {
    //     ...
    // }
}
```
```
   = note: expected type `bollard::docker::Docker<hyperlocal::client::UnixConnector>`
              found type `bollard::docker::Docker<impl hyper::client::connect::Connect>`
```

## Refactor X: embed the connection
Researching docker daemon APIs to rely on to implement my library quickly narrowed it down to a few contestants, chief among them is bollard. The other being [shiplift](https://docs.rs/shiplift/0.4.0/shiplift/)
Their model is to embed the connection into the `Docker` structure. This deprives the consumer of the library the flexibility of providing `hyper::client::connect::Connect`, at the prize of being able to consume the `Docker` structure without restrictions.

https://github.com/softprops/shiplift/blob/eb98b1916c0220e44e2d0f3c869c01a2dd037f60/src/lib.rs#L65-L69

I took the liberty to experiment with bollard to introduce this `Transport` mechanism of embedding the connection and it did not introduce significant changes nor any API breakage, save for the removal for `Docker::connect_with`.

My fork introduces the `Transport` like so:
https://github.com/veeg/bollard/blob/05ff3229d80cfc32c094f0df13af32cdefb1ef96/src/docker.rs#L102-L121

Fork available here:
https://github.com/veeg/bollard/tree/veeg/encapsulated_client

Unit and integration test pass, just like before.

*Question*: What are the argument for allowing the user to bring his own `Connect`?
Codegen? Optimization?

## Refactor Y: enum in return type
You discuss that the preferred method may be to return an enum representing the connection.
Can any of you shed any light on how this would be implemented and what problems this solution addresses? Will it solve a problem akin to mine above, where I want to provide my own library that totally abstracts bollard?

----

### Closing Remarks
After looking into the various clients available, I really prefer the API surface and completeness of `bollard`s API, compared to that of `shiplift`. However, as it stand right now, I cannot really use with the current implementation.

**Comment by fussybeaver** [comment]
Hi @veeg Thanks for the detailed write-up and the example code! I think it's a great direction.

At the moment, it's a bit cumbersome but you can make your example code work, you would need pass the hyper `Connect` trait as a type parameter, and you would rewrite your example code as follows:

```rust
use bollard::Docker;
#[cfg(unix)]
use hyperlocal::UnixConnector;
#[cfg(windows)]
use named_pipe::NamedPipeConnector;
use hyper::client::connect::Connect;

pub struct DockerTest<C> where C: Connect + Sync + 'static {
        client: Docker<C>,
}

impl<C> DockerTest<C> where C: Connect + Sync + 'static {
    pub fn local() -> DockerTest<impl Connect> {
        #[cfg(unix)]
        let docker = Docker::connect_with_unix_defaults().unwrap();
        #[cfg(windows)]
        let docker = Docker::connect_with_named_pipe_defaults().unwrap();

        DockerTest { client: docker }
    }
}

fn main() {
    let _test = DockerTest::<UnixConnector>::local();

    // test.run(|| {
    //      ...
    // }
} 
```

I think the use of a `Transport` enum is the right way to go to simplify the API a bit, as you have suggested in your fork.

To answer your question, I guess the main advantage of being able to pass in any `Connect` implementation is for mock style testing, but I guess it's fine to standardize around the yup testing library.

Do you mind opening a PR ? I would like to do a little testing over the weekend, but in principal it looks good.

**Comment by veeg** [comment]
@fussybeaver I was abit preoccupied this weekend and was unable to submit any pull request just yet. I will see to it now.

As for your counter-example, it still require the type parameter in initialization, and that is my entire point of eliminating :)

**Comment by fussybeaver** [comment]
Closed with #19
